src/MigrationActions.sol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.25;
+
+import { IERC20 }   from "lib/forge-std/src/interfaces/IERC20.sol";
+import { IERC4626 } from "lib/forge-std/src/interfaces/IERC4626.sol";
+
+interface JoinLike {
+    function vat() external view returns (VatLike);
+    function join(address, uint256) external;
+    function exit(address, uint256) external;
+}
+
+interface VatLike {
+    function hope(address) external;
+}
+
+/**
+ * @notice Actions for migrating from DAI/sDAI to NST/sNST.
+ * @dev    Also contains 1 downgrade path from NST to DAI for convenience.
+ */
+contract MigrationActions {
+
+    IERC20   public immutable dai;
+    IERC4626 public immutable sdai;
+    IERC20   public immutable nst;
+    IERC4626 public immutable snst;
+
+    VatLike  public immutable vat;
+    JoinLike public immutable daiJoin;
+    JoinLike public immutable nstJoin;
+
+    constructor(
+        address _sdai,
+        address _snst,
+        address _daiJoin,
+        address _nstJoin
+    ) {
+        sdai = IERC4626(_sdai);
+        dai  = IERC20(sdai.asset());
+        snst = IERC4626(_snst);
+        nst  = IERC20(snst.asset());
+
+        daiJoin = JoinLike(_daiJoin);
+        nstJoin = JoinLike(_nstJoin);
+        vat     = daiJoin.vat();
+
+        // Infinite approvals
+        dai.approve(_daiJoin, type(uint256).max);
+        nst.approve(_nstJoin, type(uint256).max);
+        nst.approve(_snst,    type(uint256).max);
+
+        // Vat permissioning
+        vat.hope(_daiJoin);
+        vat.hope(_nstJoin);
+    }
+
+    /**
+     * @notice Migrate `amount` of `dai` to `nst`.
+     * @param  receiver The receiver of the `nst`.
+     * @param  amount   The amount of the `dai` to migrate.
+     */
+    function migrateDAIToNST(address receiver, uint256 amount) external {
+        dai.transferFrom(msg.sender, address(this), amount);
+        daiJoin.join(address(this), amount);
+        nstJoin.exit(receiver, amount);
+    }
+
+    /**
+     * @notice Migrate `amount` of `dai` to `snst`.
+     * @param  receiver  The receiver of the `snst`.
+     * @param  amount    The amount of the `dai` to migrate.
+     * @return sharesOut The amount of `snst` shares received.
+     */
+    function migrateDAIToSNST(address receiver, uint256 amount) external returns (uint256 sharesOut) {
+        dai.transferFrom(msg.sender, address(this), amount);
+        daiJoin.join(address(this), amount);
+        nstJoin.exit(address(this), amount);
+        sharesOut = snst.deposit(amount, receiver);
+    }
+
+    /**
+     * @notice Migrate `assetsIn` of `sdai` to `nst`.
+     * @param  receiver The receiver of the `nst`.
+     * @param  assetsIn The amount of the `sdai` to migrate in assets.
+     */
+    function migrateSDAIAssetsToNST(address receiver, uint256 assetsIn) external {
```

Comment:
Not a huge fan of argument naming inconsistency, where it's `assetsIn` and `sharesIn` when sDAI enters, but simply `amount` when no DAI or NST enters. Maybe it just has to be like this, but I'd consider swapping `amount` to `assetsIn` which is technically true as well.

Reply:
changed